src/contracts/atlas/GasAccounting.sol

```diff
@@ -201,107 +227,165 @@ abstract contract GasAccounting is SafetyLocks {
 
         // Load what we can from storage so that it shows up in the gasleft() calc
-        uint256 _surcharge = cumulativeSurcharge;
-        uint256 _claims = claims;
-        uint256 _withdrawals = withdrawals;
-        uint256 _deposits = deposits;
 
-        // Remove any unused gas from the bundler's claim.
-        // TODO: consider penalizing bundler for too much unused gas (to prevent high escrow requirements for solvers)
-        uint256 gasRemainder = gasleft() * tx.gasprice;
-        gasRemainder = gasRemainder * (SURCHARGE_SCALE + SURCHARGE_RATE) / SURCHARGE_SCALE;
-        _claims -= gasRemainder;
-
-        if (_deposits < _claims + _withdrawals) {
-            // CASE: in deficit, subtract from bonded balance
-            uint256 amountOwed = _claims + _withdrawals - _deposits;
-            uint256 deficit = _assign(winningSolver, amountOwed, true, false);
-            if (deficit > 0) {
-                revert InsufficientTotalBalance(deficit);
-            }
+        uint256 _claims;
+        uint256 _writeoffs;
+        uint256 _withdrawals;
+        uint256 _deposits;
+
+        (_withdrawals, _deposits, _claims, _writeoffs, netAtlasGasSurcharge) =
+            _adjustAccountingForFees(ctx, solverGasLimit);
+
+        uint256 _amountSolverPays;
+        uint256 _amountSolverReceives;
+
+        // Calculate the balances that should be debited or credited to the solver and the bundler
+        if (_deposits < _withdrawals) {
+            _amountSolverPays += (_withdrawals - _deposits);
         } else {
-            // CASE: in surplus, add to bonded balance
-            // TODO: make sure this works w/ the surcharge 10%
-            uint256 amountCredited = _deposits - _claims - _withdrawals;
-            _credit(winningSolver, amountCredited);
+            _amountSolverReceives += (_deposits - _withdrawals);
         }
 
-        netGasSurcharge = (_claims * SURCHARGE_RATE) / (SURCHARGE_SCALE + SURCHARGE_RATE);
-
-        _claims -= netGasSurcharge;
+        // Only force solver to pay gas claims if they aren't also the bundler
+        // NOTE: If the auction isn't won, _winningSolver will be address(0).
+        if (ctx.solverSuccessful && _winningSolver != ctx.bundler) {
+            _amountSolverPays += (_claims - _writeoffs);
+            claimsPaidToBundler = (_claims - _writeoffs);
+        } else {
+            claimsPaidToBundler = 0;
+        }
 
-        cumulativeSurcharge = _surcharge + netGasSurcharge;
+        if (_amountSolverPays > _amountSolverReceives) {
+            if (!ctx.solverSuccessful) {
+                revert InsufficientTotalBalance(_amountSolverPays - _amountSolverReceives);
+            } else {
+                claimsPaidToBundler -= _assign(_winningSolver, _amountSolverPays - _amountSolverReceives, true);
+            }
+        } else {
+            _credit(_winningSolver, _amountSolverReceives - _amountSolverPays);
+        }
```

Comment:
This `else` block of code will execute if `_amountSolverPays <= _amountSolverReceives`. It will always try to `_credit()` the `winningSolver` with `_amountSolverReceives - _amountSolverPays`.

However, if there is no winning solver, it will try to `_credit()` address(0) with some amount of AtlETH. Should we handle the no winning solver case here explicitly? Maybe adding the amount to `claimsPaidToBundler` so it gets sent to the bundler instead?

Reply:
great catch.

Fix: 62e2131